README.md
CardFlight iOS SDK Library
=================

Introduction
------------

The CardFlight iOS SDK is used to process card present and card not-present transactions in your iOS application.

CardFlight's SDK is based around keeping it as simple as possible while keeping the highest level of [security](https://developers.cardflight.com/help/security) at the forefront of all that we do. Take out the pain of PCI-compliance when building your app.

Authentication is done through your API Keys and processing is done through the Account Tokens. All connections to CardFlight's API are done through HTTPS over HSTS.

Setup
----------

### CocoaPods Installation

[CocoaPods](http://www.cocoapods.org/) is the recommended method of installing CardFlight.

##### Podfile

```ruby
platform :ios, '8.0'
pod "CardFlight"
```

### Manual Installation

Add the **libCardFlightLibrary.a** and all header files into your project. Add the following frameworks to you project:

- AVFoundation.framework
- AudioToolbox.framework
- MediaPlayer.framework
- MessageUI.framework
- CoreGraphics.framework
- ExternalAccessory.framework
- libstdc++.6.0.9.dylib

### Logging

The SDK includes a logging mode to assist with debugging that will print out statements in the console. Be sure to disable it in the production code.

```
- (void)setLogging:(BOOL)logging;
```

### Initialize

Add **CardFlight.h** to your AppDelegate and set the API Key and Account Token.

##### Example

```
[[CFTSessionManager sharedInstance] setApiToken:@"e9cb15260f08e438b782952895d4ba4d"
accountToken:@"acc_04ff8bf654afb265"];
```

The CardFlight SDK is broken into easy-to-manage components. You just include the ones that you want to use in the header files of the classes that need to access those components.

SDK Documentation
--------------

- [Errors](https://developers.cardflight.com/docs/api#errors)
- [Reader Initialization](https://developers.cardflight.com/docs/api#reader_initialization)
- [Swipe Card](https://developers.cardflight.com/docs/api#swipe_card)
- [Swipe Response](https://developers.cardflight.com/docs/api#swipe_card_response)
- [Keyed Entry](https://developers.cardflight.com/docs/api#keyed_entry)
- [Keyed Response](https://developers.cardflight.com/docs/api#keyed_response)
- [Charge Card](https://developers.cardflight.com/docs/api#process_payment)
- [Refund Charge](https://developers.cardflight.com/docs/api#refund_charge)


Supported Platforms
-----------------------

All CardFlight readers work on **iOS version 8.0 or above**. Our SDK supports a wide array of iOS platforms. [Click here](https://developers.cardflight.com/docs/ios) to view an updated list.


Looking for Android?
-----------------

We've got you covered. [Click here](https://github.com/CardFlight/cardflight-android) to learn more about our CardFlight Android SDK.


CHANGELOG.md
### 3.5.1
* Remove support for Attache reader
* Fixed transaction result not being returned on cancelled EMV transaction  (issue 64)
* Improved messaging on prematurely disconnected A200 reader (issue 47)
* Voids and Refunds now accept callback URLs
* Added ability to start contact transaction on B550 with card inserted
* Fix vault feature
* Added support for receipt data from a CFTCharge
* Removed subspecs from podspec

### 3.5
* Added support for Bluetooth readers
* Add ability to charge a vaulted card
* Wrapped up commlink error from A200 (issue 42)
* Fix for crash on signature upload (issue 56)
* Added error for no audio permissions in the host app
* Updated enumerations
* Updated keyed entry view touch target

### 3.2.1
* Fix for EMV signature method
* Restored functionality to deprecated setAPIKey method (issue 46)
* Deprecated attache reader class. Support for this reader type will be removed in a future release. Please contact support@cardflight.com if you need an alternative solution

### 3.2
* Fixed Card Removed error (issue 32)
* Added completion callback after setting API Key and Account Token (issue 34)
* Fixed EMV Messaging (issue 35, 38)
* Improved EMV Reader connection logic (issue 36)

### 3.1
* Cleaned up compiler warnings
* Added destroy reader method

### 3.0
* Added EMV support
* Added iOS 9 support
* Removed iOS 6 support
* Added bitcode support

### 2.0.6
* Update to expiration date

### 2.0.5
* Update to EMV backend
* Fix for last4 when card number is blank
* Added ability to save log statements to a file

### 2.0.4
* Fix for multiple transactions in a row

### 2.0.2
* Improved reader connection reliability

### 2.0.2
* Documentation clarification
* Improved reader connection times

### 2.0.1
* Bug fix for beginSwipe method

### 2.0
* Added support for multiple readers
* Added new transaction method
* Added new CFTReader delegate methods
* Removed all previously deprecated methods

### 1.9.1
* Added fix for connecting to iPhone 6/6 Plus

### 1.9
* Added methods for auth and capture
* Added optional zip code to manual entry

### 1.8.3
* Additional fix for file corruption

### 1.8.2
* Fix for file corruption

### 1.8.1
* Fixed Attache reader supspec

### 1.8
* Added 64-bit support

### 1.7.1
* Fixed bug with refund amount, now accepts amounts in dollar and cents

### 1.7.0
* Added metadata to charges
* Created new refund method

### 1.6.3
* Added ability to copy and serialize CFTCharge
* Improved deprecation warnings
* Updated documentation

### 1.6.2
* Exposed encrypted swipe data

### 1.6.1
* Created separate static libraries for shuttle and attache readers

### 1.6
* Changed to subspec for Attache reader class (pod 'CardFlight/Attache')
* Changed CardFlight class to CFTSessionManager
* Added Void function

### 1.5.0
* Changed custom manual entry to unified field
* Added manual entry protocol
* Deprecated old version of manual entry
* Added support for NSCopying of CFTCard
* Change readerDelegate to CFTReaderDelegate

### 1.4.0
* Added 64 bit support
* Added flag for attache reader
* Deprecated connect and disconnect methods

### 1.3.0
* Added name field for custom manual entry
* Added text field delegate forwarders for custom manual entry
* Bugfixes

### 1.2

* Added logging mode for debugging
* Added becomeFirstResponder forwarder
* Added textFieldShouldReturn protocol method

### 1.1.3

* Added callback for user cancelled swipe
* Fixed callback for reader timeout
* Fixed ability to change account token

### 1.1.2

* Fixed url encoding of transaction descriptions

### 1.1.1

* Added support for reading swipes from generic cards
* Fixed tokenization card errors

### 1.1.0

* Added support for new readers
* Added support for timeout duration

### 1.0.2

* Fixes bug with errors returning nil

### 1.0.1

* Minor bug fixes

### 1.0.0

* Initial release


CFTEnum.h
/*
 *****************************************************************
 * CFTEnum.h
 *
 * Copyright (c) 2015 CardFlight Inc. All rights reserved.
 *****************************************************************
 */

typedef NS_ENUM(NSUInteger, CFTEMVMessage) {
    EMVMessage_UNKNOWN = 0,
    EMVMessage_AMOUNT_OK = 1,
    EMVMessage_APPLICATION_BLOCKED = 2,
    EMVMessage_APPROVED = 3,
    EMVMessage_CALL_BANK = 4,
    EMVMessage_CAPK_FAIL = 5,
    EMVMessage_CARD_BLOCKED = 6,
    EMVMessage_CARD_ERROR = 7,
    EMVMessage_CARD_NOT_SUPPORTED = 8,
    EMVMessage_CLEAR_DISPLAY = 9,
    EMVMEssage_CONDITIONS_NOT_SATISFIED = 10,
    EMVMessage_DECLINED = 11,
    EMVMessage_DEVICE_ERROR = 12,
    EMVMessage_DIP_OR_SWIPE = 13,
    EMVMessage_ENTER_AMOUNT = 14,
    EMVMessage_ENTER_PIN = 15,
    EMVMessage_ICC_CARD_REMOVED = 16,
    EMVMessage_INCORRECT_PIN = 17,
    EMVMessage_INSERT_CARD = 18,
    EMVMessage_INVALID_ICC_DATA = 19,
    EMVMessage_MISSING_DATA = 20,
    EMVMessage_NO_EMV_APP = 21,
    EMVMessage_NOT_ACCEPTED = 22,
    EMVMessage_NOT_ICC = 23,
    EMVMessage_ONLINE_REQUIRED = 24,
    EMVMessage_PIN_OK = 25,
    EMVMessage_PLEASE_WAIT = 26,
    EMVMessage_PRESENT_ONLY_ONE_CARD = 27,
    EMVMessage_PROCESSING = 28,
    EMVMessage_PROCESSING_ERROR = 29,
    EMVMessage_REMOVE_CARD = 30,
    EMVMessage_FALLBACK = 31,
    EMVMessage_SET_AMOUNT_CANCEL_OR_TIMEOUT = 32,
    EMVMessage_TERMINATED = 33,
    EMVMessage_TRANSACTION_CANCELLED = 34,
    EMVMessage_TRY_AGAIN = 35,
    EMVMessage_TRY_TAP_AGAIN = 36,
    EMVMessage_TRY_DIP_AGAIN = 37,
    EMVMessage_TRY_SWIPE_AGAIN = 38,
    EMVMessage_TAP_DIP_OR_SWIPE = 39,
    EMVMessage_USE_CHIP_READER = 40,
    EMVMessage_USE_MAG_READER = 41,
    EMVMessage_WELCOME = 42,
    EMVMessage_NO_MESSAGE = 43,
    EMVMessage_NO_PERMISSION = 44
};

typedef NS_ENUM(NSUInteger, CFTEMVCardState) {
    CFTEMVCardState_UNKNOWN = 0,
    CFTEMVCardState_INSERTED = 1,
    CFTEMVCardState_WAITING_FOR_REMOVAL = 2,
    CFTEMVCardState_REMOVED = 3,
    CFTEMVCardState_TAPPED = 4
};

typedef NS_ENUM (NSUInteger, CFTCardType) {
    CFTCardTypeVisa = 0,
    CFTCardTypeMasterCard = 1,
    CFTCardTypeAmex = 2,
    CFTCardTypeDiscover = 3,
    CFTCardTypeJCB = 4,
    CFTCardTypeDinersClub = 5,
    CFTCardTypeUnknown = 6
};

typedef NS_ENUM (NSUInteger, CFTReaderState) {
    ReaderState_UNKNOWN = 0,
    ReaderState_WAITING_FOR_CONNECT = 1,
    ReaderState_ATTACHED = 2,
    ReaderState_CONNECTED = 3,
    ReaderState_WAITING_FOR_TAP = 4,
    ReaderState_WAITING_FOR_DIP = 5,
    ReaderState_WAITING_FOR_SWIPE = 6,
    ReaderState_SWIPE_DETECTED = 7,
    ReaderState_CARD_INSERTED = 8,
    ReaderState_REMOVE_CARD = 9,
    ReaderState_DISCONNECTED = 10,
    ReaderState_NOT_FOUND = 11,
    ReaderState_NO_PERMISSION = 12
};

typedef NS_ENUM(NSUInteger, CFTReaderType) {
    CFTReaderType_UNKNOWN = 0,
    CFTReaderType_SHUTTLE = 1,
    CFTReaderType_BTMAG = 2,
    CFTReaderType_A100 = 3,
    CFTReaderType_A200 = 4,
    CFTReaderType_B550 = 5,
    CFTReaderType_B500 = 6,
    CFTReaderType_A250 = 7,
    CFTReaderType_B200 = 8,
    CFTReaderType_B250 = 9
};
